File: parallel.py
from mpi4py import MPI
import numpy as np
import matplotlib.pyplot as plt

def parallel_k_means(X, k, max_iters=100):
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    # Master initializes data and centroids
    if rank == 0:
        # ...code to generate or load data X if needed...
        centroids = X[np.random.choice(X.shape[0], k, replace=False)]
        print(f"[Rank {rank}] Initial centroids:\n", centroids)
    else:
        X = None
        centroids = None

    # Broadcast data dimensions and centroids
    centroids = comm.bcast(centroids, root=0)

    # Scatter the data among processes
    num_points = 0 if rank != 0 else X.shape[0]
    num_points = comm.bcast(num_points, root=0)
    local_data_size = num_points // size
    remainder = num_points % size
    counts = [local_data_size + 1 if p < remainder else local_data_size for p in range(size)]
    starts = [sum(counts[:p]) for p in range(size)]
    local_data = np.zeros((counts[rank], 2), dtype='float64')  # adjust shape
    if rank == 0:
        for p in range(size):
            if p != 0:
                comm.Send(X[starts[p]:starts[p]+counts[p]], dest=p, tag=77)
        local_data = X[starts[0]:starts[0]+counts[0]]
    else:
        comm.Recv(local_data, source=0, tag=77)

    for iteration in range(max_iters):
        # Compute cluster assignments locally
        distances = np.zeros((local_data.shape[0], k))
        for i, c in enumerate(centroids):
            distances[:, i] = np.linalg.norm(local_data - c, axis=1)
        labels_local = np.argmin(distances, axis=1)

        # Partial sums for each cluster
        partial_sums = np.zeros((k, local_data.shape[1]))
        counts_local = np.zeros(k, dtype=int)
        for idx, label in enumerate(labels_local):
            partial_sums[label] += local_data[idx]
            counts_local[label] += 1

        # Reduce to get global sums and counts
        global_sums = np.zeros_like(partial_sums)
        global_counts = np.zeros_like(counts_local)
        comm.Reduce(partial_sums, global_sums, op=MPI.SUM, root=0)
        comm.Reduce(counts_local, global_counts, op=MPI.SUM, root=0)

        # Master updates centroids
        if rank == 0:
            for i in range(k):
                if global_counts[i] > 0:
                    centroids[i] = global_sums[i] / global_counts[i]
            print(f"[Rank {rank}] Iteration {iteration} updated centroids:\n", centroids)

        # Broadcast new centroids
        centroids = comm.bcast(centroids, root=0)

    # Gather labels from each process
    all_labels = None
    comm.Barrier()
    labels_local = labels_local.astype(np.int32)  # ensure int32
    if rank == 0:
        all_labels = np.empty(num_points, dtype=np.int32)
    comm.Gatherv(labels_local, (all_labels, counts, starts, MPI.INT), root=0)

    if rank == 0:
        print("[Rank 0] Final centroids:\n", centroids)
        print("[Rank 0] Final labels:\n", all_labels)
        plt.scatter(X[:, 0], X[:, 1], c=all_labels)
        plt.scatter(centroids[:, 0], centroids[:, 1], c='red', marker='x')
        plt.title("Final clustering result")
        plt.show()
    return centroids

if __name__ == "__main__":
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    if rank == 0:
        np.random.seed(42)
        X = np.vstack((np.random.randn(100, 2) + np.array([4, 4]),
                       np.random.randn(100, 2) + np.array([-4, -4]),
                       np.random.randn(100, 2)))
    else:
        X = None

    final_centroids = parallel_k_means(X, k=3, max_iters=10)

File: parallel.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// ...existing code for allocate_2d, free_2d, or generate_data...
// (Refer to sequential.c for those helpers if needed)

void compute_distances(double** X, double** centroids, double* distances,
                       int local_count, int k, int dim) {
    // ...existing code logic adapted...
    for(int i = 0; i < local_count; i++) {
        for(int j = 0; j < k; j++) {
            double sum = 0.0;
            // ...existing code to compute distance...
            distances[i*k + j] = sqrt(sum);
        }
    }
}

void assign_clusters(double* distances, int* labels_local,
                     int local_count, int k) {
    // ...existing code logic adapted...
}

void update_centroids(double** sums, int* counts, double** centroids,
                      int k, int dim) {
    for(int i = 0; i < k; i++) {
        if(counts[i] > 0) {
            for(int d = 0; d < dim; d++) {
                centroids[i][d] = sums[i][d] / counts[i];
            }
        }
    }
}

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    int n = 600, dim = 2, k = 3, max_iters = 100;
    double** X = NULL;
    int* labels = NULL;

    if(rank == 0) {
        // ...existing code to allocate and generate data...
        // e.g. X = allocate_2d(n, dim); generate_data(X, n, dim);
        labels = (int*)malloc(n * sizeof(int));
    }

    // Distribute data counts
    int base_count = n / size;
    int remainder = n % size;
    int* counts = (int*)malloc(size * sizeof(int));
    int* displs = (int*)malloc(size * sizeof(int));
    for(int i = 0; i < size; i++) {
        counts[i] = base_count + (i < remainder ? 1 : 0);
    }
    displs[0] = 0;
    for(int i = 1; i < size; i++) {
        displs[i] = displs[i-1] + counts[i-1];
    }

    // Allocate local data
    double** X_local = NULL;
    int local_count = counts[rank];
    // ...allocate X_local with local_count rows...

    // Prepare contiguous buffer for scatter/gather
    double* sendbuf = NULL;
    double* recvbuf = (double*)malloc(local_count * dim * sizeof(double));
    if(rank == 0) {
        sendbuf = (double*)malloc(n * dim * sizeof(double));
        // Flatten X into sendbuf
        // ...existing code to copy X into sendbuf...
    }

    // Scatterv data
    int* sendcounts = (int*)malloc(size*sizeof(int));
    int* senddispls = (int*)malloc(size*sizeof(int));
    for(int i = 0; i < size; i++) {
        sendcounts[i] = counts[i]*dim;
        senddispls[i] = displs[i]*dim;
    }
    MPI_Scatterv(sendbuf, sendcounts, senddispls, MPI_DOUBLE,
                 recvbuf, local_count*dim, MPI_DOUBLE,
                 0, MPI_COMM_WORLD);

    // Rebuild X_local from recvbuf
    // ...existing code to map recvbuf to X_local...

    // Initialize centroids on rank 0
    double** centroids = NULL;
    if(rank == 0) {
        centroids = /* allocate_2d(k, dim) */;
        // ...existing code to pick random initial centroids...
    } 
    else {
        // ...allocate for centroids...
    }
    // Broadcast centroids
    for(int i = 0; i < k; i++) {
        MPI_Bcast(centroids[i], dim, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    }

    // Iteration loop
    double* distances = (double*)malloc(local_count*k*sizeof(double));
    int* labels_local = (int*)malloc(local_count*sizeof(int));
    for(int iter = 0; iter < max_iters; iter++) {
        // Store old centroids for convergence
        double* old = (double*)malloc(k*dim*sizeof(double));
        // ...copy centroids to old...

        compute_distances(X_local, centroids, distances,
                          local_count, k, dim);
        assign_clusters(distances, labels_local, local_count, k);

        // Partial sums
        double** partial_sums = /* allocate_2d(k, dim) */;
        int* local_counts = (int*)calloc(k, sizeof(int));
        // ...accumulate partial sums for each cluster...

        // Reduce sums and counts on rank 0
        double** global_sums = NULL;
        int* global_counts = NULL;
        if(rank == 0) {
            global_sums = /* allocate_2d(k, dim) */;
            global_counts = (int*)calloc(k, sizeof(int));
        }
        for(int i = 0; i < k; i++) {
            MPI_Reduce(partial_sums[i], (rank==0?global_sums[i]:NULL),
                       dim, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
            MPI_Reduce(&local_counts[i], (rank==0?&global_counts[i]:NULL),
                       1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
        }

        // Rank 0 updates centroids
        if(rank == 0) {
            update_centroids(global_sums, global_counts, centroids, k, dim);
            // Check convergence
            int converged = 1;
            // ...compare centroids with old...
            if(converged) {
                // ...free local memory...
                // broadcast break signal if desired
            }
        }

        // Broadcast updated centroids
        for(int i = 0; i < k; i++) {
            MPI_Bcast(centroids[i], dim, MPI_DOUBLE, 0, MPI_COMM_WORLD);
        }
        // ...free local temporary allocations...
    }

    // Gather labels
    int* all_labels = NULL;
    if(rank == 0) {
        all_labels = (int*)malloc(n*sizeof(int));
    }
    MPI_Gatherv(labels_local, local_count, MPI_INT,
                all_labels, counts, displs, MPI_INT,
                0, MPI_COMM_WORLD);

    if(rank == 0) {
        // ...print final centroids, or handle results...
        // free_2d(X); free(labels);
        free(all_labels);
    }
    // ...cleanup local arrays, etc...

    MPI_Finalize();
    return 0;
}
